### PR TITLE
[proxy] add analytics endpoint

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

[Internal context](https://gitpod.slack.com/archives/C047WL9JBLG/p1680019749050709)

- This PR introduces a new `/analytics` endpoint for collecting analytics from different clients in a centralized manner. The endpoint utilizes basic authentication from segment to determine whether to proxy to trusted or untrusted sources.
- In preview environments, Gitpod components (server, ws-manager-bridge, ws-proxy) are configured to use this endpoint with a trusted key, and the same configuration will be applied to production environments in a subsequent PR.
- In workspace services (supervisor and Gitpod CLI) are also reconfigured to use this endpoint but without a key in preview environments (i.e., untrusted), and the same configuration to be applied to the production environment in a subsequent PR.
- After deployment, local clients like VS Code Desktop will also be updated to use this endpoint.
- This PR as well fixes analytics installer options for GH actions.

## Related Issue(s)

Fixes #

## How to test

- Start a new workspace and check that analytics is configured now based on segment endpoint not key.
Screenshot 2023-04-12 at 21 39 29

- Run gp tasks list command and then check staging trusted that gp_command event is registered there.
- Check that events from other Gitpod components are registered as well, i.e. workspace started and stopped, and so on.

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
- [x] analytics=segment

/hold

### CODE REVIEW

**Comment by roboquat** [comment]
@akosyakov: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by filiptronicek** [comment]
Would rate limiting be in-scope for this PR? Maybe it would be useful if we are going to implement the reverse proxy anyway

**Comment by akosyakov** [comment]
> Would rate limiting be in-scope for this PR? Maybe it would be useful if we are going to implement the reverse proxy anyway

Was not planing to till we see actual abuse.

File: components/proxy/plugins/analytics/analytics.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		MaxIdleConns:          0,
+		MaxIdleConnsPerHost:   100,
+		IdleConnTimeout:       30 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+
+	a.reverseProxy = http.StripPrefix("/analytics", reverseProxy)
```
Comment (akosyakov) [inline components/proxy/plugins/analytics/analytics.go:93]: probably it is better to add `/analytics/segment` route that we can add other routes later if necessary, not only proxy to segment

**Comment by easyCZ** [comment]
@akosyakov given our discussion yesterday, is this something you're looking to land as is? If not, could we move to draft until it's eligible for review?

**Comment by akosyakov** [comment]
> @akosyakov given our discussion yesterday, is this something you're looking to land as is? If not, could we move to draft until it's eligible for review?

Refactored it since yesterday, but I will need your input. Now `/analytics` endpoint accept from all clients, and decide whether to delegate to trusted or untrusted source based on segment basic auth, i.e. if client is using trusted key then it goes to trusted, if no key or untrusted then to untrusted.

File: components/common-go/analytics/analytics.go [hunk 2]
```diff
@@ -45,7 +45,14 @@ func NewFromEnvironment() Writer {
 		return &logAnalyticsWriter{}
 	case "segment":
 		log.Debug("segment analytics")
-		return &segmentAnalyticsWriter{Client: segment.New(os.Getenv("GITPOD_ANALYTICS_SEGMENT_KEY"))}
+		client, err := segment.NewWithConfig(os.Getenv("GITPOD_ANALYTICS_SEGMENT_KEY"), segment.Config{
+			Endpoint: os.Getenv("GITPOD_ANALYTICS_SEGMENT_ENDPOINT"),
+		})
+		if err != nil {
+			log.WithError(err).Error("cannot create segment client")
```
Comment (easyCZ) [inline components/common-go/analytics/analytics.go:52]: How critical are analytics? Do we want to exit/propagate error if we can't set it up?

File: components/proxy/plugins/analytics/analytics.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	}
+	if a.untrustedProxy != nil && (segmentKey == "" || segmentKey == a.untrustedProxy.segmentKey) {
+		if segmentKey == "" {
+			r.SetBasicAuth(a.untrustedProxy.segmentKey, "")
+		}
+		a.untrustedProxy.ServeHTTP(w, r)
+		return nil
+	}
+	return next.ServeHTTP(w, r)
+}
```
Comment (easyCZ) [inline components/proxy/plugins/analytics/analytics.go:122]: Would be nice to have a test for this

File: install/installer/pkg/components/proxy/deployment.go [hunk 4]
```diff
@@ -258,6 +264,12 @@ func deployment(ctx *common.RenderContext) ([]runtime.Object, error) {
 								}, {
 									Name:  "WORKSPACE_HANDLER_FILE",
 									Value: strings.ToLower(string(ctx.Config.Kind)),
+								}, {
+									Name:  "ANALYTICS_PLUGIN_TRUSTED_SEGMENT_KEY",
+									Value: trustedSegmentKey,
+								}, {
+									Name:  "ANALYTICS_PLUGIN_UNTRUSTED_SEGMENT_KEY",
+									Value: untrustedSegmentKey,
 								}},
```
Comment (easyCZ) [inline install/installer/pkg/components/proxy/deployment.go:273]: Can we pull this into something like `common.AnalyticsEnv` similar to other re-usable parts?

File: install/installer/pkg/config/v1/config.go [hunk 5]
```diff
@@ -180,8 +180,9 @@ type Observability struct {
 }
 
 type Analytics struct {
-	SegmentKey string `json:"segmentKey"`
-	Writer     string `json:"writer"`
+	SegmentKey      string `json:"segmentKey"`
+	Writer          string `json:"writer"`
```
Comment (easyCZ) [inline install/installer/pkg/config/v1/config.go:184]: Out of curiosity, what does Writer mean?

File: components/proxy/plugins/analytics/analytics.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+func (a *Analytics) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.Handler) error {
+	segmentKey, _, ok := r.BasicAuth()
+	if !ok {
+		return next.ServeHTTP(w, r)
+	}
+	if a.trustedProxy != nil && segmentKey == a.trustedProxy.segmentKey {
+		a.trustedProxy.ServeHTTP(w, r)
+		return nil
+	}
+	if a.untrustedProxy != nil && (segmentKey == "" || segmentKey == a.untrustedProxy.segmentKey) {
```
Comment (easyCZ) [inline components/proxy/plugins/analytics/analytics.go:114]:
```suggestion
	shouldProxytoUntrusted := a.untrustedProxy != nil && (segmentKey == "" || segmentKey == a.untrustedProxy.segmentKey)
	if  shouldProxyToUntrusted {
```
Mostly for ease of legibility as the expression can be a mouthful

File: components/proxy/plugins/analytics/analytics.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+	trustedSegmentKey := os.Getenv("ANALYTICS_PLUGIN_TRUSTED_SEGMENT_KEY")
+	if trustedSegmentKey != "" {
+		a.trustedProxy = newSegmentProxy(segmentEndponit, errorLog, trustedSegmentKey)
+	}
+
+	return nil
+}
+
+func newSegmentProxy(segmentEndponit *url.URL, errorLog *log.Logger, segmentKey string) *segmentProxy {
```
Comment (easyCZ) [inline components/proxy/plugins/analytics/analytics.go:81]:
```suggestion
func newSegmentProxy(segmentEndpoint *url.URL, errorLog *log.Logger, segmentKey string) *segmentProxy {
```

**Review by easyCZ — COMMENTED** [review]
In general looks good

File: install/installer/pkg/config/v1/config.go (hunk 5 above)
Comment (akosyakov) [reply install/installer/pkg/config/v1/config.go:184]: i think here it is none, log or segment

File: components/common-go/analytics/analytics.go (hunk 2 above)
Comment (akosyakov) [reply components/common-go/analytics/analytics.go:52]: I cannot answer for everybody. In workspace it is not super critical to interrupt user flow, i.e. if we don't know about usage of gp cli it is not the end of the world. Not sure about other components like dashboard.

File: install/installer/pkg/components/proxy/deployment.go (hunk 4 above)
Comment (akosyakov) [reply install/installer/pkg/components/proxy/deployment.go:273]: Maybe I misunderstand, but I think this PR already adds such config: https://github.com/gitpod-io/gitpod/pull/17162/files#diff-590a17333c6831c6ff532edbab0b818ee3c49fb033d163722552bbbfafd31a8cR466

Comment (easyCZ) [reply install/installer/pkg/components/proxy/deployment.go:273]: You're right, sorry. Missed that.

**Comment by akosyakov** [comment]
I've addressed comments and added tests. I will test more manually later. @mads-hartmann maybe you can have a look whether it is something what you think would help you with dedicated

**Review by mads-hartmann — COMMENTED** [review]
I have two questions related to dedicated

1. Initially we'll want the analytics endpoint to just drop all events in Dedicated. I couldn't quite tell if it's currently possible to have the analytics caddy plugin drop events right now or we'd need to modify it to achieve that.
2. In the near future we'll want to (1) sanitise events (2) send them to a Firehose Delivery Stream instead of Segment. It looks like Caddy plugins are powerful enough that we'll be able to do that, is that your understanding too?

**Comment by akosyakov** [comment]
> Initially we'll want the analytics endpoint to just drop all events in Dedicated. I couldn't quite tell if it's currently possible to have the analytics caddy plugin drop events right now or we'd need to modify it to achieve that.

if you don't configure any segment keys, analytics plugin will drop everything.

> In the near future we'll want to (1) sanitise events (2) send them to a Firehose Delivery Stream instead of Segment. It looks like Caddy plugins are powerful enough that we'll be able to do that, is that your understanding too?

yes, it is go program, you have access to request object to modify the body before forwarding somewhere

**Review by mads-hartmann — APPROVED** [review]
Based on [this comment](https://github.com/gitpod-io/gitpod/pull/17162#issuecomment-1508211032) this PR looks good to me! 🛹

**Comment by akosyakov** [comment]
I added tests for cases when keys are not configured.

**Review by easyCZ — APPROVED** [review]

**Pull request merged**